src/CodeGen_C.cpp

```diff
@@ -181,7 +179,38 @@ const string preamble =
     "}\n";
 }
 
-CodeGen_C::CodeGen_C(ostream &s) : IRPrinter(s), id("$$ BAD ID $$") {}
+CodeGen_C::CodeGen_C(ostream &s, bool is_header, const std::string &guard) : IRPrinter(s), id("$$ BAD ID $$"), is_header(is_header) {
+    if (is_header) {
+        // If it's a header, emit an include guard.
+        stream << "#ifndef HALIDE_" << print_name(guard) << '\n'
+               << "#define HALIDE_" << print_name(guard) << '\n';
+    }
+
+    stream << "#ifdef __cplusplus\n";
+    stream << "#define HALIDE_EXTERN_C extern \"C\"\n";
+    stream << "#else\n";
+    stream << "#define HALIDE_EXTERN_C\n";
```

Comment:
It adds a small bit of complexity to do it this way (we can't put extern "C" around *everything* that CodeGen_C produces), I'll change it anyways though. I don't feel strongly about it. FWIW, this change is irrelevant of this PR, I just happened to notice that we were unconditionally generating `extern "C"` (i.e. compiling with a true C compiler would break) while changing this code around.